btree/btree.c

```diff
@@ -4820,15 +4827,34 @@ static void vkvv_make(struct slot *slot)
 
 static void vkvv_val_resize(struct slot *slot, int vsize_diff)
 {
-	struct vkvv_head *h     = vkvv_data(slot->s_node);
-	struct dir_rec   *rec   = vkvv_get_dir_addr(slot->s_node);
-	int               idx   = slot->s_idx;
-	int               count = h->vkvv_used;
+	struct vkvv_head *h              = vkvv_data(slot->s_node);
+	struct dir_rec   *rec            = vkvv_get_dir_addr(slot->s_node);
+	int               idx            = slot->s_idx;
+	int               count          = h->vkvv_used;
+	void             *end_val_addr   = vkvv_val(slot->s_node, count);
+	uint32_t          dir_size       = sizeof(struct dir_rec) * (count + 1);
+	void             *start_dir_addr = (void*)rec;
+	void             *end_dir_addr   = start_dir_addr + dir_size;
+	void             *end_key_addr   = vkvv_key(slot->s_node, count);
 	void             *start_val_addr;
 	uint32_t          total_vsize;
+	int               diff;
 
 	M0_PRE(slot->s_idx < h->vkvv_used && h->vkvv_used > 0);
 
+	if (vsize_diff > 0 &&
+	    (end_val_addr - end_dir_addr) < vsize_diff) {
+		diff = vsize_diff - (end_val_addr - end_dir_addr);
+
+		if (start_dir_addr - end_key_addr < diff ||
+		    (end_val_addr - end_dir_addr) < 0)
```

Comment:
Why do you need this 2nd condition (end_val_addr - end_dir_addr) < 0  ??

Reply:
This particular condition was used for testing. I will remove it in the upcoming update.